### DISCUSSION TITLE
I'm forever looking to see a headless CMS that can content store to Git.

### DISCUSSION THREAD

**Discussion by paul-hammant** [payloadcms/payload#2459] [Feature Requests & Ideas]
https://github.com/prose/prose is close, and Netlify has a proprietary option. 

I'm not against a database aspect of CMS, but just think for certain classes of application the DB can amount to a cache/index of something that is canonically stored in source control.

Sure, "store to git" isn't a cinch. Github and Gitlab has APIs. I've done work in the past to hack 'kitchen sink' APIs for various in order to showcase the idea of vcs-backed storage.

**Comment by teunmooij** [comment]
Why? I would not recommend it.

But if you really want that, https://github.com/payloadcms/payload is the CMS you should use. You'd only need to write some simple hooks to accomplish just that.
And you could also write it as a plugin and share it for others to use.

**Comment by miguderp** [comment]
You might be looking for [Tina CMS](https://tina.io/)

**Reply by qbeauperin** [reply]
Yup, or [Decap CMS](https://github.com/decaporg/decap-cms) (previously called Netlify CMS - which was not and still isn't proprietary but open source and free), or [Static CMS](https://github.com/StaticJsCMS/static-cms) (an interesting an actively worked on fork of Decap/Netlify CMS).

**Comment by paul-hammant** [comment]
I've tried Tina previously, thanks @miguderp. Thanks for the info on the rename @qbeauperin and the fork.

**Comment by tyteen4a03** [comment]
Bumping this - Tina seems to be unmaintained and has a lot of shortcomings compared to Payload.

**Reply by GermanJablo** [reply]
They had a brief period of inactivity, but they are now active again.
By the way, there is also Keystatic.

**Reply by linkb15** [reply]
Yea, Keystatic is so similar with Sanity and Payload in a sense they build the Schema using the Code - which is really good developer experience with the Hot Reload and seeing your changes take effect quickly. 

Feature wise, there are limitations with Git-based CMS in general - like you only can do a simple query of the file systems, the rest you have to calculate yourself (wish there is more better way of doing that) and running on edge runtime (runtime that does not support file system), you would utilize to Webpack / Vite import glob to read folders, and parse the data yourself.

**Reply by tyteen4a03** [reply]
In scenarios where it is more important for the machine to be able to read the data than APIs (such as game configurations), Git-based is the much better choice.

**Comment by jmikrut** [comment]
Well, because we have a database adapter pattern now, we could technically build an adapter for git storage - although I think that will have lots of challenges, specifically regarding how deeply you can query in payload. It would be interesting to explore this idea. It’s probably not going to be for sometime if explored from the payload team ourselves, but if someone wanted to take an attempt at building the adapter, we would definitely help!

**Comment by victorantos** [comment]
On the querying tradeoff raised above: with Git as the store you don't query at read time, you precompute. The thing I built for this (my project, so weigh accordingly) rebuilds the site and a set of static JSON indexes on every commit -- /api/site.json plus per-collection index files -- so a "query" is a CDN fetch of a file that already holds the answer. Filter by tag, list by date, search: fine. Arbitrary joins, user-generated content, anything read-then-write: not fine, and for those Payload with a real database is the honest answer rather than bending Git into one.

https://github.com/plain-cms/plain -- MIT, 14 stars, one maintainer. Early, and the bus factor is one.
